### DISCUSSION TITLE
Save answers for questions in customer accounts

### DISCUSSION THREAD

**Discussion by Sohalt** [pretix/pretix#2741] [Ideas] [answered]
Some repeated events always ask the same questions on every event (e.g. think of an event that provides a free T-Shirt for participants, asking for the T-Shirt size or an event asking for a short descriptive text to be printed alongside the participant's name on their badge).

It would be nice to be able to save answers to these questions in a customer account, so you don't have to enter the same information every time.

Usually answers to questions in the checkout process are stored as part of the individual event and get deleted/anonymized when using the data shredder, after the event is over. Storing answers with the customer account in plain text for use in future events somewhat defeats the purpose of a data shredder. To circumvent this problem, I propose to store the answers to questions in an encrypted format. The decryption key should be derived from the user's password (or rather an intermediate key, which itself is encrypted using the user's password, so a password change does not require reencrypting all answers, but simply reencrypting the interediate key). This allows the user to access their prior answers during the checkout process. The (potentially modified) answers then get copied to/stored in the specific event in plain text (as is currently the case with all "new" answers), so they are visible to the organizers. After the event is over, the data can be shredded as usual and the answers are still available to the user in the checkout phase for the next event.

The actual decryption of old answers could even happen entirely in the user's webbrowser, so the server never sees old answers, which might have already been shredded, but even if the decryption happens on the server, it is still better than keeping the answers completely in plain text, because it prevents a data leak in the case the database gets stolen (only the currently unencrypted, not-shredded answers are accessible).

**Comment by raphaelm** [comment, marked as answer]
> It would be nice to be able to save answers to these questions in a customer account, so you don't have to enter the same information every time.

This is already implemented and ready to be used!

> Usually answers to questions in the checkout process are stored as part of the individual event and get deleted/anonymized when using the data shredder, after the event is over. Storing answers with the customer account in plain text for use in future events somewhat defeats the purpose of a data shredder.

I disagree. Saving question answers to the customer account is optional and opt-in. So there is clear and active consent of the user to store the data beyond the single event.

I think an encryption scheme like you mention is way more complex than necessary here, especially given the fact that the data most often is not very sensitive. Also, encrypting with the user password is pretty inconvenient because people forget their password all the time, **especially** their password for an online shop they only use once a year.

**Reply by Sohalt** [reply]
> This is already implemented and ready to be used!

Oh, sorry, I thought you could only save contact information.

> I disagree. Saving question answers to the customer account is optional and opt-in. So there is clear and active consent of the user to store the data beyond the single event.

Is there a way to easily delete this data in the future, without having to delete one's customer account?

**Reply by raphaelm** [reply]
> Oh, sorry, I thought you could only save contact information.

No problem :) 

It should show up like this once you're logged in:

![2022-07-30-204348_1395x401_scrot](https://user-images.githubusercontent.com/64280/181937305-13098870-fd1e-40e6-8628-812e97a757a5.png)

> Is there a way to easily delete this data in the future, without having to delete one's customer account?

Yup :) 

![2022-07-30-204404_1344x226_scrot](https://user-images.githubusercontent.com/64280/181937316-2a425790-6ba9-49f0-bc4e-c5e783d44a1e.png)
